librad/src/net/tls.rs

```diff
@@ -15,157 +15,156 @@
+    cfg.cert_resolver = Arc::new(CertResolver::new(signer, cert));
+
+    Ok(cfg)
+}
 
-    cfg
+struct Cert {
+    expires: Date,
+    cert: rustls::Certificate,
 }
 
-/// Extract the [`PeerId`] from a certificate as generated by [`gen_cert`].
-pub fn extract_peer_id(cert_der: &[u8]) -> ASN1Result<PeerId> {
-    yasna::parse_der(&cert_der, |reader| {
-        reader.read_sequence(|reader| {
-            let pk = reader.next().read_sequence(|reader| {
-                // X.509 version
-                reader.next().read_der()?;
-                // serial number
-                reader.next().read_der()?;
-                // signature algorithm
-                reader.next().read_der()?;
-                // issuer
-                reader.next().read_der()?;
-                // validity
-                reader.next().read_der()?;
-                // subject
-                reader.next().read_der()?;
-
-                // THE MEAT: subjectPublicKeyInfo
-                let pk = reader.next().read_sequence(|reader| {
-                    // subject key algorithm
-                    reader.next().read_der()?;
-                    // here we go
-                    let (value, bits) = reader.next().read_bitvec_bytes()?;
-                    // check for overflow
-                    if bits % 8 == 0 {
-                        Ok(value)
-                    } else {
-                        Err(ASN1Error::new(ASN1ErrorKind::Invalid))
-                    }
-                })?;
-
-                // Must consume the rest (extensions)
-                reader.next().read_der()?;
-
-                Ok(pk)
-            })?;
-
-            let peer_id = PublicKey::from_slice(&pk)
-                .map(PeerId::from)
-                .ok_or_else(|| ASN1Error::new(ASN1ErrorKind::Invalid))?;
-
-            // Consume the rest
-            // signatureAlgorithm
-            reader.next().read_der()?;
-            // signature
-            reader.next().read_der()?;
-
-            Ok(peer_id)
+struct CertResolver {
+    signer: BoxedSigner,
+    cert: RwLock<Cert>,
+}
+
+impl CertResolver {
+    fn new<S>(signer: S, cert: x509::Certificate) -> Self
+    where
+        S: Signer + Clone + Send + Sync + 'static,
+        S::Error: std::error::Error + Send + Sync + 'static,
+    {
+        let signer = BoxedSigner::from(SomeSigner { signer });
+        let expires = x509::validity_time_as_date(&cert.tbs_certificate.validity.not_after);
+        let cert = rustls::Certificate(cert.to_der());
+        Self {
+            signer,
+            cert: RwLock::new(Cert { expires, cert }),
+        }
+    }
+
+    /// Get the certificate, or generate a fresh one if it has expired
+    fn cert(&self) -> Result<rustls::Certificate, BoxedSignError> {
+        let now = OffsetDateTime::now_utc().date();
+        let read = self.cert.read().unwrap();
+        if now >= read.expires {
+            drop(read);
+            let fresh = x509::Certificate::generate(&self.signer)?;
```

Comment:
@kim I presume this where you mean its fatal? Sounds reasonable to me to panic here. It would mean something is _terrrrrribly_ wrong.

Reply:
I went for logging the error -- the signer can fail all over the place, and it could be intermittent